packages/react/tooltip/src/Tooltip.tsx

```diff
@@ -61,109 +128,76 @@ interface TooltipProps {
   const [trigger, setTrigger] = React.useState<HTMLButtonElement | null>(null);
   const contentId = useId();
+  const openTimerRef = React.useRef(0);
+  const delayDuration = delayDurationProp ?? context.delayDuration;
+  const openDelay = context.isOpenDelayed ? delayDuration : 0;
+  const wasOpenDelayed = usePrevious(Boolean(openDelay));
+  const { onTooltipOpen, onTooltipClose } = context;
   const [open = false, setOpen] = useControllableState({
     prop: openProp,
     defaultProp: defaultOpen,
     onChange: onOpenChange,
   });
-  const [stateAttribute, setStateAttribute] = React.useState<StateAttribute>(
-    openProp ? 'instant-open' : 'closed'
-  );
 
-  // control open state using state machine subscription
-  React.useEffect(() => {
-    const unsubscribe = stateMachine.subscribe(({ state, context }) => {
-      if (state === 'open' && context.id === contentId) {
-        setOpen(true);
-      } else {
-        setOpen(false);
-      }
-    });
+  const stateAttribute = React.useMemo(() => {
+    return open ? (wasOpenDelayed ? 'delayed-open' : 'instant-open') : 'closed';
+  }, [wasOpenDelayed, open]);
 
-    return unsubscribe;
-  }, [contentId, setOpen]);
-
-  // sync state attribute with using state machine subscription
-  React.useEffect(() => {
-    const unsubscribe = stateMachine.subscribe(({ state, context }) => {
-      if (context.id === contentId) {
-        if (state === 'open') {
-          setStateAttribute(context.delayed ? 'delayed-open' : 'instant-open');
-        } else {
-          setStateAttribute('closed');
-        }
-      } else {
-        setStateAttribute('closed');
-      }
+  const handleOpen = React.useCallback(() => {
+    setOpen((prevOpen) => {
+      if (!prevOpen) document.dispatchEvent(new CustomEvent(TOOLTIP_OPEN));
+      return true;
     });
+    onTooltipOpen();
+  }, [setOpen, onTooltipOpen]);
 
-    return unsubscribe;
-  }, [contentId]);
-
-  const handleFocus = React.useCallback(
-    () => stateMachine.send({ type: 'FOCUS', id: contentId }),
-    [contentId]
-  );
-  const handleOpen = React.useCallback(
-    () => stateMachine.send({ type: 'OPEN', id: contentId, delayDuration }),
-    [contentId, delayDuration]
-  );
-  const handleClose = useCallbackRef(() => {
-    stateMachine.send({ type: 'CLOSE', id: contentId, skipDelayDuration });
-  });
-
-  // send transition if the component unmounts
-  React.useEffect(() => () => handleClose(), [handleClose]);
-
-  // if we're controlling the component
-  // put the state machine in the appropriate state
-  useLayoutEffect(() => {
-    if (openProp === true) {
-      stateMachine.send({ type: 'OPEN', id: contentId });
-    }
-  }, [contentId, openProp]);
+  React.useEffect(() => {
+    return () => window.clearTimeout(openTimerRef.current);
+  }, []);
 
   return (
     <PopperPrimitive.Root {...popperScope}>
-      <TooltipProvider
-        scope={__scopeTooltip}
+      <TooltipImplProvider
```

Comment:
Why is this called `TooltipImplProvider` when there are no `TooltipImpl`?

Reply:
I can't call this `TooltipProvider` because there is a `TooltipProvider` part now. Not sure what to name it? The `TooltipProvider` is the provider for all tooltips and this provider is the provider for **each** tooltip.